### PR TITLE
first version of cloud-init, incomplete but put framework here for discussion.

### PR DESCRIPTION
**Pull request by YinYangOfDao**
etcd, kube api/proxy/weave and app pod like mysql can init on infra. 
next need to check UI and job submission

### CODE REVIEW

File: src/ClusterBootstrap/v2deploy.sh [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# !/bin/bash
+<<<<<<< HEAD
```
Comment (xudifsd) [inline src/ClusterBootstrap/v2deploy.sh]: conflict not solved.

File: src/ClusterBootstrap/cloud_init_deploy.py [hunk 2]
```diff
@@ -1,105 +1,121 @@
...
-    "postmasterdeploymentscript" : (["platform-scripts"], lambda x: get_platform_script_directory(x)+"post-master-deploy.sh"),
-    "mastercleanupscript" : (["platform-scripts"], lambda x: get_platform_script_directory(x)+"cleanup-master.sh"),
-    "masterdeploymentlist" : (["platform-scripts"], lambda x: get_platform_script_directory(x)+"deploy.list"),
-    #worker deployment scripts
-    "preworkerdeploymentscript" : (["platform-scripts"], lambda x: get_platform_script_directory(x)+"pre-worker-deploy.sh"),
-    "postworkerdeploymentscript" : (["platform-scripts"], lambda x: get_platform_script_directory(x)+"post-worker-deploy.sh"),
-    "workercleanupscript" : (["platform-scripts"], lambda x: get_platform_script_directory(x)+"cleanup-worker.sh"),
-    "workerdeploymentlist" : (["platform-scripts"], lambda x: get_platform_script_directory(x)+"deploy.list"),
-    }
+    "dockerprefix": (["cluster_name"], lambda x: x.lower()+"/"),
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/cloud_init_deploy.py]: add space before and after `+`

File: src/ClusterBootstrap/cloud_init_deploy.py [hunk 3]
```diff
@@ -163,7 +185,9 @@ def load_node_list_by_role_from_config(config, roles):
         Nodes += temp_nodes
     return Nodes, config
 
-# Get the list of nodes for a particular service    
+# Get the list of nodes for a particular service
+
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/cloud_init_deploy.py]: remove two empty lines

File: src/ClusterBootstrap/cloud_init_deploy.py [hunk 4]
```diff
@@ -194,166 +220,815 @@ def get_node_lists_for_service(service, config):
...
-        config["prometheus"]["host"] = None if len(get_node_lists_for_service("prometheus", config))==0 \
-                                        else get_node_lists_for_service("prometheus", config)[0]
+        config["prometheus"]["host"] = None if len(get_node_lists_for_service("prometheus", config)) == 0 \
+            else get_node_lists_for_service("prometheus", config)[0]
     config = update_docker_image_config(config)
 
-    config["api_servers"] = "https://"+config["kubernetes_master_node"][0]+":"+str(config["k8sAPIport"])
-    config["etcd_endpoints"] = ",".join(["https://"+x+":"+config["etcd3port1"] for x in config["etcd_node"]])
+    config["api_servers"] = "https://" + \
+        config["kubernetes_master_node"][0]+":"+str(config["k8sAPIport"])
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/cloud_init_deploy.py]: space before and after `+`

File: src/ClusterBootstrap/cloud_init_deploy.py [hunk 5]
```diff
@@ -194,166 +220,815 @@ def get_node_lists_for_service(service, config):
...
+    config["apiserver_ssl_ip"] = "\n".join(["IP.{} = {}".format(i, sslip) for i, sslip in enumerate(
+        [config["api-server-ip"]] + config["ssl_localhost_ips"] + masterips)])
+
+    # kube-apiserver aggregator use easyrsa to generate crt files, we need to generate a group of master names for it.
+    # It does not care if it's a DNS name or IP.
+    masternames = []
+    for i, value in enumerate(config["kubernetes_master_node"]):
+        masternames.append(value)
+    config["apiserver_names_ssl_aggregator"] = ",".join(
+        ["DNS:"+name for i, name in enumerate(masternames)])
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/cloud_init_deploy.py]: space before and after `+`

Comment (Anbang-Hu) [reply src/ClusterBootstrap/cloud_init_deploy.py]: same for line 378

File: src/ClusterBootstrap/cloud_init_deploy.py [hunk 6]
```diff
@@ -194,166 +220,815 @@ def get_node_lists_for_service(service, config):
...
+            launch_order_file = os.path.join(dirname, "launch_order")
+            if os.path.isfile(launch_order_file):
+                servicedic[service] = launch_order_file
+                with open(launch_order_file, 'r') as f:
+                    allservices = f.readlines()
+                    for filename in reversed(allservices):
+                        filename = filename.strip()
+                        filename = os.path.join(dirname, filename)
+                        if os.path.isfile(filename):
+                            servicedic[service+"/"+os.path.splitext(
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/cloud_init_deploy.py]: space before and after `+`

File: src/ClusterBootstrap/cloud_init_deploy.py (hunk 2 above)
Comment (YinYangOfDao) [reply src/ClusterBootstrap/cloud_init_deploy.py]: would add in next PR.

File: src/ClusterBootstrap/cloud_init_deploy.py (hunk 3 above)
Comment (YinYangOfDao) [reply src/ClusterBootstrap/cloud_init_deploy.py]: remove in next PR

File: src/ClusterBootstrap/scripts/cloud_init_infra.sh [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    cntr=2
+    until kubectl create -f $svc ; do
+        sleep 5;
+        cntr=$((cntr-1))
+        echo "waiting for ${svc}, ${cntr} more attempts"
+        if [ "$cntr" -le 0 ]; then
+            break
+        fi
+    done
+
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/scripts/cloud_init_infra.sh]: remove empty line

File: src/utils/ConfigUtils.py [hunk 8]
```diff
@@ -3,26 +3,30 @@
...
+import yaml
 
 verbose = False
 
+
 def expand_path(path):
     return expanduser(path)
 
 # Test if a certain Config entry exist
+
```
Comment (Anbang-Hu) [inline src/utils/ConfigUtils.py]: Remove two empty lines

File: src/utils/ConfigUtils.py [hunk 9]
```diff
@@ -3,26 +3,30 @@
...
+
 def fetch_dictionary(dic, entry):
     if isinstance(entry, list):
-        # print "Fetch " + str(dic) + "@" + str(entry) + "==" + str( dic[entry[0]] ) 
-        if isinstance( dic, list ):
+        # print "Fetch " + str(dic) + "@" + str(entry) + "==" + str( dic[entry[0]] )
```
Comment (Anbang-Hu) [inline src/utils/ConfigUtils.py]: Remove commented code

File: src/utils/ConfigUtils.py (hunk 8 above)
Comment (Anbang-Hu) [reply src/utils/ConfigUtils.py]: same for 39 - 40, 45 - 46

**Pull request merged**